### PR TITLE
[Sema] Define library levels to detect public import of private modules and more

### PR DESCRIPTION
**Pull request by xymus**
Intro the concept of library access/distribution level that indicates if it will be distributed widely (`api`) or reserved for use inside the organization (`spi`).

In practice, setting the flag `-library-level api` will warn on public imports of SPI modules and suggests using `@_implementationOnly`. The imported module SPI-ness is determined from the location it was loaded from and from whether it belongs to a private clang modulemap.( rdar://62934005)

Future work:
- The flag `-library-level spi` should affect type-checking, it allows the use of SPI decls in API as the whole module is considered SPI. (rdar://75335462)
- Infer the library level in the build system from the installation path.
- Preserve the library level value in the serialized swiftmodule and use it to determine the SPI-ness of imported Swift modules. Let's wait for more compilers to understand this flag in order to avoid a condfail.
- Support the `ipi` level for project-internal modules?

### CODE REVIEW

**Comment by xymus** [comment]
@swift-ci Please smoke test

**Comment by xymus** [comment]
I forgot to commit a test.

@swift-ci Please smoke test

**Comment by xymus** [comment]
I may have to wait before I can land the last commit *The flag -library-level spi should affect type-checking, it allows the use of SPI decls in API as the whole module is considered SPI. (rdar://75335462)* as this one requires to print the flag in the swift interface which will may lead to condfails.

**Comment by xymus** [comment]
@swift-ci Please smoke test

File: lib/AST/Module.cpp
Comment (nkcsgexi) [inline lib/AST/Module.cpp:2264]: Nit: start with `llvm::Twine(ctx.SearchPathOpts.SDKPath)` here.

**Review by nkcsgexi — COMMENTED** [review]
This is a really nice approach! Any reason we don't emit `-library-level` into the interface so we don't need to infer the level from search paths?

File: include/swift/Basic/LangOptions.h
Comment (xymus) [reply include/swift/Basic/LangOptions.h:69]: Using API as the default would mean that all projects get warnings on imports of private modules until they manually set the flag `-library-level spi` or we get a build system to set it automatically. I'm really tempted by that to get results faster but the rollout will be smoother if we wait for the build system to infer the library level.

Also, `Other` is now relevant to IPI/build time modules. We don't see much of those in Swift but we do import IPI clangmodules.

**Comment by xymus** [comment]
@nkcsgexi It's on my todo list but we should wait a bit to print `-library-level` to the interfaces to avoid a condfail. It's not automatic so it wouldn't be an instant condfail but I'm worried project owners would have to hold off using this feature because of it.

**Review by nkcsgexi — APPROVED** [review]
It makes sense. Thank you for the explanation!

File: lib/AST/Module.cpp
Comment (xymus) [reply lib/AST/Module.cpp]: While applying this I came across `llvm::sys::path::append` which simplifies greatly this piece of code.

**Comment by xymus** [comment]
@swift-ci Please smoke test

**Comment by xymus** [comment]
@swift-ci Please smoke test macOS

**Comment by xymus** [comment]
@swift-ci Please smoke test macOS

**Pull request merged**
